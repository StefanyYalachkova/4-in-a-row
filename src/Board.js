import React, { useEffect, useState } from "react";
import { Grid } from '@mui/material';
import { MoveBall } from "./MoveBall";
import { SVGBallPlayer1 } from "./SVGBallPlayer1";
import { SVGBallPlayer2 } from "./SVGBallPlayer2";
import { getIsBallObjectEmpty } from "./utils";

const Board = (props) => {

    let rowSize = props.boardSize.row;
    let colSize = props.boardSize.col;

    const initialValue = () => {
        const matrixOfObjects = [];

        for (let i = 0; i < rowSize; i++) {
            matrixOfObjects[i] = [];
            for (let j = 0; j < colSize; j++) {
                matrixOfObjects[i][j] = ({ isPlayerOneBall: false, isPlayerTwoBall: false });
            };
        };

        return matrixOfObjects;
    };

    const [board, setBoard] = useState(initialValue());

    useEffect(() => {
        if (props.restartGame) {
            setBoard(initialValue());
        } else {
            props.runOnUpdate(board);
        };
    }, [board, props.restartGame]);

    const displayBoard = (boardState, keyPrefix) => {
        let result = [];
        const rowCount = boardState.length;
        const colCount = boardState[0].length;

        let getColElement = (keyPrefix, rowIndex, col) => {
            return (
                <Grid className={'cow-container'} item xs={1} key={`${keyPrefix}-${rowIndex}-${col}`}>
                    <svg
                        viewBox="0 0 24 24"
                    >
                        <rect width="30" height="30" style={{ fill: "rgb(0,0,255)", strokeWidth: 0, stroke: "rgb(0,0,0)" }} />
                        {displayBall(boardState[rowIndex][col])}
                    </svg>
                </Grid>
            );
        };

        const getColumns = (keyPrefix, rowIndex) => {
            const columns = [];
            for (let col = 0; col < colCount; col++) {
                columns.push(getColElement(keyPrefix, rowIndex, col));
            };

            return columns;
        };

        let getRowElement = (key, rowIndex) => (
            <Grid className={'row-container'} xs={9} key={key}>
                {getColumns(keyPrefix, rowIndex)}
            </Grid>
        );

        for (let row = 0; row < rowCount; row++) {
            result.push(getRowElement(`${keyPrefix}-${row}`, row));
        };

        return result;
    };

    const displayBall = (element) => {
        if (element.isPlayerOneBall || element.isPlayerTwoBall) {

            return (
                element.isPlayerOneBall === true
                    ? <SVGBallPlayer1 />
                    : <SVGBallPlayer2 />
            );
        } else {

            return (
                <svg>
                    <circle
                        cx="12" cy="12" r="8"
                        strokeWidth="6" stroke="white"
                        fill="white"
                    />
                </svg >
            );
        };
    };

    const moveDownBall = (position, activeElement) => {
        const copy = [...board];
        let isTurnFinished = false;

        let startLine = copy.length - 1;
        let valueToSet = { isPlayerOneBall: false, isPlayerTwoBall: false };

        while (startLine >= 0 && !isTurnFinished) {
            if (getIsBallObjectEmpty(copy[startLine][position])) {
                if (activeElement.isPlayerOneBall === true) {
                    valueToSet = { isPlayerOneBall: true, isPlayerTwoBall: false };
                    copy[startLine][position] = valueToSet;
                } else {
                    valueToSet = { isPlayerOneBall: false, isPlayerTwoBall: true };
                    copy[startLine][position] = valueToSet;
                }
                isTurnFinished = true;
            } else {
                startLine--;
            };
        };

        setBoard(copy);
    };

    const isLastCellFull = (position) => {
        const { isPlayerOneBall, isPlayerTwoBall } = board[0][position];

        return isPlayerOneBall || isPlayerTwoBall;
    };

    const playground = () => {

        return (
            <div>
                <MoveBall
                    displayBoard={displayBoard}
                    handleMoveDown={moveDownBall}
                    handleIsLastCellFull={isLastCellFull}

                    shouldStop={props.shouldReset}
                    shouldRestartGame={props.restartGame}
                    afterGameRestart={props.setRestartGame}
                    winnerBall={props.winnerBall}
                    rowSize={rowSize}
                    colSize={colSize}
                    ref={props.moveBallReff}
                />
                <Grid>
                    {displayBoard(board, 'board')}
                </Grid>
            </div>
        );
    };

    return playground();
};

export { Board };