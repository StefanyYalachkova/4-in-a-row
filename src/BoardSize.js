import React from "react";
import { Button, TextField } from '@mui/material';

const BoardSize = (props) => {

    const handleConfirm = (event) => {
        event.preventDefault();
        props.onClick();
    };

    return (
        <div>
            <form>
                <h1>Set board size :</h1>
                <span id="rowsField">Rows:</span>
                <TextField
                    type="number"
                    name="rows"
                    onChange={props.setSize}
                    helperText={props.isRowNumberValid ? '' : <span id="errorMessage">Enter a number greater than 4.</span>}
                />
                <span id="colsField">Columns:</span>
                <TextField
                    type="number"
                    name="columns"
                    onChange={props.setSize}
                    helperText={props.isColNumberValid ? '' : <span id="errorMessage">Enter a number greater than 4.</span>}
                />
                <span id="nInARowField">N in a row:</span>
                <TextField
                    type="number"
                    name="nInARows"
                    onChange={props.setSize}
                    helperText={props.isNinARowValid ? '' : <span id="errorMessage">Enter a number greater than 2.</span>}
                />
                <Button disabled={props.shouldDisabledButtonBoardSize} variant="contained" type="submit" onClick={handleConfirm}> Confirm </Button>
            </form>
        </div>
    );
};

export { BoardSize };