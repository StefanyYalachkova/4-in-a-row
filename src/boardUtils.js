import { Grid } from "@mui/material";
import { SVGBallPlayer1 } from "./SVGBallPlayer1";
import { SVGBallPlayer2 } from "./SVGBallPlayer2";

const getIsBallObjectEmpty = (ballObject) => {
    return !(ballObject.isPlayerOneBall || ballObject.isPlayerTwoBall);
};

const areBallObjectsEqual = (ballObject1, ballObject2) => {
    return ballObject1.isPlayerOneBall === ballObject2.isPlayerOneBall
        && ballObject1.isPlayerTwoBall === ballObject2.isPlayerTwoBall;
};

const getBoardInitialState = (rowSize, colSize) => {
    const matrixOfObjects = [];

    for (let i = 0; i < rowSize; i++) {
        matrixOfObjects[i] = [];
        for (let j = 0; j < colSize; j++) {
            matrixOfObjects[i][j] = ({ isPlayerOneBall: false, isPlayerTwoBall: false });
        };
    };

    return matrixOfObjects;
};

const displayBoard = (boardState, keyPrefix) => {
    let result = [];
    const rowCount = boardState.length;
    const colCount = boardState[0].length;

    let getColElement = (keyPrefix, rowIndex, col) => {
        return (
            <Grid className={'cow-container'} item xs={1} key={`${keyPrefix}-${rowIndex}-${col}`}>
                <svg viewBox="0 0 24 24">
                    <rect width="30" height="30" style={{ fill: "rgb(0,0,255)", strokeWidth: 0, stroke: "rgb(0,0,0)" }} />
                    <DisplayBall element={boardState[rowIndex][col]} />
                </svg>
            </Grid>
        );
    };

    const getColumns = (keyPrefix, rowIndex) => {
        const columns = [];

        for (let col = 0; col < colCount; col++) {
            columns.push(getColElement(keyPrefix, rowIndex, col));
        };

        return columns;
    };

    const getRowElement = (key, rowIndex) => (
        <Grid className={'row-container'} xs={9} key={key}>
            {getColumns(keyPrefix, rowIndex)}
        </Grid>
    );

    for (let row = 0; row < rowCount; row++) {
        result.push(getRowElement(`${keyPrefix}-${row}`, row));
    };

    return result;
};

const DisplayBall = (props) => {
    const { element } = props;

    let result = (
        <svg>
            <circle
                cx="12" cy="12" r="8"
                strokeWidth="6" stroke="white"
                fill="white"
            />
        </svg>
    );

    if (element.isPlayerOneBall || element.isPlayerTwoBall) {
        result = (
            element.isPlayerOneBall
                ? <SVGBallPlayer1 />
                : <SVGBallPlayer2 />
        );
    };

    return result;
};

export { getIsBallObjectEmpty, areBallObjectsEqual, getBoardInitialState, displayBoard, DisplayBall };
