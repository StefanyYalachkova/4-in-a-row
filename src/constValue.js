const RIGHT_ARROW = 39;
const LEFT_ARROW = 37;
const DOWN_ARROW = 40;
const A_KEYCODE = 65;
const S_KEYCODE = 83;
const D_KEYCODE = 68;

const MAX_SCORE = 3;

const PLAYER_ONE = 'playerOne';
const PLAYER_TWO = 'playerTwo';

const INITIAL_ROW_NUMBER = 8;
const INITIAL_COL_NUMBER = 8;
const INITIAL_WINCOUNT_NUMBER = 4;

const MIN_VALID_NAME_LENGTH = 3;
const MIN_VALID_ROW_COL_NUMBER = 4;
const MIN_VALID_WINCOUNT_NUMBER = 2;


export {
    RIGHT_ARROW, LEFT_ARROW, DOWN_ARROW, A_KEYCODE, S_KEYCODE, D_KEYCODE,
    MAX_SCORE,
    PLAYER_ONE, PLAYER_TWO,
    INITIAL_ROW_NUMBER, INITIAL_COL_NUMBER, INITIAL_WINCOUNT_NUMBER,
    MIN_VALID_NAME_LENGTH, MIN_VALID_ROW_COL_NUMBER, MIN_VALID_WINCOUNT_NUMBER
};
