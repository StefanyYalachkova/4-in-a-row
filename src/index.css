body {
  padding: 20px;
  background-color: #f1d2e0e1;
}

button[type=submit] {
  font-family: sans-serif;
  background-color: rgb(0, 0, 255);
  padding: 10px;
  border-radius: 8px;
  height: 100%;
  cursor: pointer;
  margin: 5px;
  position: relative;
  left: 5%;
}

form,
h1 {
  text-align: center;
  padding: 15px;
}

h2 {
  text-align: inherit;
}

.playerOne-container {
  display: flex;
  align-items: center;
  justify-content: right;
}

.playerTwo-container {
  display: flex;
  align-items: center;
  justify-content: left;
}

.board-container {
  align-items: center;
  justify-content: center;
}

.row-container {
  display: flex;
  margin: -5px;
  align-items: center;
  justify-content: flex-end
}

.balls-container {
  display: flex;
  border: solid;
  align-items: center;
  justify-content: center;
}

.balls {
  width: auto;
  height: auto;

}

.firstLine-container {
  padding-bottom: 8px;
  align-items: center;
  justify-content: center;
  border-style: hidden;
  outline: none;
}

.winner-container {
  text-align: right;
  padding: 50px;
}

.errorMessage {
  color: red;
  font-size: 15px;
}

#playerField,
#rowsField,
#colsField,
#winCountField {
  padding: 20px;
  font-weight: bold;
  font-size: 20px;
}

.player-container {
  text-align: center;
  font-size: xx-large;
}
