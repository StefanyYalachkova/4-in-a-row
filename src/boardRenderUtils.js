import { areBallObjectsEqual, getIsBallObjectEmpty } from "./boardUtils";

const getWinnerForThisRound = (board = [[]], additionalParams = {}) => {
    const { playerOneName, playerTwoName, winCount } = additionalParams;

    const roundWinner = getWinner(board, winCount);
    let winnerForThisRound = '';

    if (roundWinner.isPlayerOneBall) {
        winnerForThisRound = playerOneName;
    } else if (roundWinner.isPlayerTwoBall) {
        winnerForThisRound = playerTwoName;
    };

    return winnerForThisRound;
};

const getWinner = (boardSize, winCount) => {

    const roundWinner = (getVerticalWinnerCell(boardSize, winCount)
        || getHorizontalWinnerCell(boardSize, winCount)
        || getRightDiagonalWinnerCell(boardSize, winCount)
        || getLeftDiagonalWinnerCell(boardSize, winCount)
        || {});

    return roundWinner;
};

const initialPosition = (boardSize, winCount) => {
    return Math.max((Math.ceil(boardSize.length / 2) - 1), winCount - 1);
};

const getVerticalWinnerCell = (boardSize, winCount) => {
    let result = null;

    for (let row = initialPosition(boardSize, winCount); row < boardSize.length; row++) {
        for (let col = 0; col < boardSize[row].length; col++) {
            if (!getIsBallObjectEmpty(boardSize[row][col])) {
                let hasWinningChance = true;
                let i = 1;

                while (hasWinningChance && i < winCount) {
                    hasWinningChance = areBallObjectsEqual(boardSize[row][col], boardSize[row - i][col])
                    i++;
                };

                if (hasWinningChance) {
                    result = boardSize[row][col];
                };
            };
        };
    };

    return result;
};

const getHorizontalWinnerCell = (boardSize, winCount) => {
    let result = null;

    for (let row = 0; row < boardSize.length; row++) {
        for (let col = 0; col < (boardSize[row].length - winCount + 1); col++) {
            if (!getIsBallObjectEmpty(boardSize[row][col])) {
                let hasWinningChance = true;
                let i = 1;

                while (hasWinningChance && i < winCount) {
                    hasWinningChance = areBallObjectsEqual(boardSize[row][col], boardSize[row][col + i]);
                    i++;
                };

                if (hasWinningChance) {
                    result = boardSize[row][col];
                };
            };
        };
    };

    return result;
};

const getRightDiagonalWinnerCell = (boardSize, winCount) => {
    let result = null;

    for (let row = initialPosition(boardSize, winCount); row < boardSize.length; row++) {
        for (let col = 0; col < (boardSize[row].length - winCount + 1); col++) {
            if (!getIsBallObjectEmpty(boardSize[row][col])) {
                let hasWinningChance = true;
                let i = 1;

                while (hasWinningChance && i < winCount) {
                    hasWinningChance = areBallObjectsEqual(boardSize[row][col], boardSize[row - i][col + i]);
                    i++;
                };

                if (hasWinningChance) {
                    result = boardSize[row][col];
                };
            };
        };
    };

    return result;
};

const getLeftDiagonalWinnerCell = (boardSize, winCount) => {
    let result = null;

    for (let row = initialPosition(boardSize, winCount); row < boardSize.length; row++) {
        for (let col = initialPosition(boardSize, winCount); col < boardSize[row].length; col++) {
            if (!getIsBallObjectEmpty(boardSize[row][col])) {
                let hasWinningChance = true;
                let i = 1;

                while (hasWinningChance && i < winCount) {
                    hasWinningChance = areBallObjectsEqual(boardSize[row][col], boardSize[row - i][col - i]);
                    i++;
                };

                if (hasWinningChance) {
                    result = boardSize[row][col];
                };
            };
        };
    };

    return result;
};

export { getWinnerForThisRound };
