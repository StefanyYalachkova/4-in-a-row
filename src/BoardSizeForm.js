import React from "react";
import { Button, TextField } from '@mui/material';
import { MIN_VALID_ROW_COL_NUMBER, MIN_VALID_WINCOUNT_NUMBER } from "./constValue";

const BoardSizeForm = (props) => {
    const { setSize, isRowNumberValid, isColNumberValid, isWinCountValid, winCount, rowSize, colSize } = props;
    const shouldDisableBoardSizeButton = rowSize < MIN_VALID_ROW_COL_NUMBER || colSize < MIN_VALID_ROW_COL_NUMBER || winCount < MIN_VALID_WINCOUNT_NUMBER;

    const handleConfirm = (event) => {
        event.preventDefault();
        props.onClick();
    };

    return (
        <form>
            <h1>Set board size :</h1>
            <span id="rowsField">Rows:</span>
            <TextField
                type="number"
                name="rows"
                onChange={setSize}
                helperText={isRowNumberValid ? '' : <span className="errorMessage">Enter a number greater than 4.</span>}
            />
            <span id="colsField">Columns:</span>
            <TextField
                type="number"
                name="columns"
                onChange={setSize}
                helperText={isColNumberValid ? '' : <span className="errorMessage">Enter a number greater than 4.</span>}
            />
            <span id="winCountField">N in a row:</span>
            <TextField
                type="number"
                name="winCount"
                onChange={setSize}
                helperText={isWinCountValid ? '' : <span className="errorMessage">Enter a number greater than 2.</span>}
            />
            <Button disabled={shouldDisableBoardSizeButton} variant="contained" type="submit" onClick={handleConfirm}> Confirm </Button>
        </form>
    );
};

export { BoardSizeForm };
