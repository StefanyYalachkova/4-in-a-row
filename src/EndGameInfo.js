import { Button, Grid } from "@mui/material";
import React, { useEffect } from "react";
import { MAX_SCORE } from "./constValue";
import { PlayerScore } from "./PlayerScore";

const EndGameInfo = (props) => {
    const { moveBallRef, winner, setViewBoard, renderNewRound } = props;
    const { playerOne, playerTwo, setPlayerOne, setPlayerTwo } = props.playersProps;

    useEffect(() => {
        if (playerOne.score === MAX_SCORE || playerTwo.score === MAX_SCORE) {
            setViewBoard(false)
        }
    }, [renderNewRound]);

    const getWinnerMessage = (newFocusRef) => {
        let result = null;

        if (winner) {
            result = (
                <div>
                    <b>{winner} wins!</b>
                    <Button variant="contained" type="submit" onClick={() => renderNewRound(newFocusRef, winner)}> New Round </Button>
                </div>
            );
        };

        return result;
    };

    const gameOverInfo = () => {
        let displayResult = '';

        if (playerOne.score === MAX_SCORE || playerTwo.score === MAX_SCORE) {
            displayResult = (
                <Grid container >
                    <Grid className={'player-container'} item xs={7}>
                        <PlayerScore
                            player={playerOne}
                            setPlayer={setPlayerOne}
                        />
                    </Grid>
                    <Grid item xs={1}></Grid>
                    <Grid className={'player-container'} >
                        <PlayerScore
                            player={playerTwo}
                            setPlayer={setPlayerTwo}
                        />
                    </Grid>
                </Grid>
            );
        } else {
            displayResult = getWinnerMessage(moveBallRef);
        };

        return displayResult;
    };

    return (
        <div>
            {gameOverInfo()}
        </div>
    );
};

export { EndGameInfo };
