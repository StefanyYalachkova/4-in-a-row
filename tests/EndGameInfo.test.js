import { fireEvent, render, screen } from "@testing-library/react";
import React from "react";
import renderer from 'react-test-renderer';
import { EndGameInfo } from "../src/EndGameInfo";

describe('EndGameInfo test', () => {
    it('renders correctly', () => {
        const playerOne = { name: 'Stef', score: 0 };
        const playerTwo = { name: 'Angel', score: 0 };
        const setPlayerOne = jest.fn();
        const setPlayerTwo = jest.fn();
        const moveBallRef = jest.fn();
        const winner = playerOne.name;
        const setViewBoard = jest.fn();
        const renderNewRound = jest.fn();

        const tree = renderer
            .create(<EndGameInfo
                playersProps={{ playerOne, playerTwo, setPlayerOne, setPlayerTwo }}
                moveBallRef={moveBallRef}
                winner={winner}
                setViewBoard={setViewBoard}
                renderNewRound={renderNewRound}
            />)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
    it('renders correctly gameOverInfo', () => {
        const playerOne = { name: 'Stef', score: 3 };
        const playerTwo = { name: 'Angel', score: 0 };
        const setPlayerOne = jest.fn();
        const setPlayerTwo = jest.fn();
        const moveBallRef = jest.fn();
        const winner = playerOne.name;
        const setViewBoard = jest.fn();
        const renderNewRound = jest.fn();

        const tree = renderer
            .create(<EndGameInfo
                playersProps={{ playerOne, playerTwo, setPlayerOne, setPlayerTwo }}
                moveBallRef={moveBallRef}
                winner={winner}
                setViewBoard={setViewBoard}
                renderNewRound={renderNewRound}
            />)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
    it('handles New Round button correctly', () => {

        const playerOne = { name: 'Stef', score: 1 };
        const playerTwo = { name: 'Angel', score: 0 };
        const setPlayerOne = jest.fn();
        const setPlayerTwo = jest.fn();
        const moveBallRef = jest.fn();
        const winner = playerOne.name;
        const setViewBoard = jest.fn();
        const renderNewRound = jest.fn();
        const newFocusRef = jest.fn();

        render(<EndGameInfo
            playersProps={{ playerOne, playerTwo, setPlayerOne, setPlayerTwo }}
            moveBallRef={moveBallRef}
            winner={winner}
            setViewBoard={setViewBoard}
            renderNewRound={renderNewRound}
            onClick={() => renderNewRound(newFocusRef, winner)}
        />)
        fireEvent.click(screen.getByText('New Round'))
        expect(renderNewRound).toBeCalled();
    });
});
