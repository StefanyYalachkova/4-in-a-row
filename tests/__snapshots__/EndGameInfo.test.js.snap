// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EndGameInfo test renders correctly 1`] = `
<div>
  <div>
    <b>
      Stef
       wins!
    </b>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="submit"
    >
       New Round 
    </button>
  </div>
</div>
`;

exports[`EndGameInfo test renders correctly gameOverInfo 1`] = `
<div>
  <div
    className="MuiGrid-root MuiGrid-container css-11lq3yg-MuiGrid-root"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-7 player-container css-uqwprf-MuiGrid-root"
    >
      <div>
        <h2>
          Stef
          : 
          3
        </h2>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 css-u78q51-MuiGrid-root"
    />
    <div
      className="MuiGrid-root player-container css-vj1n65-MuiGrid-root"
    >
      <div>
        <h2>
          Angel
          : 
          0
        </h2>
      </div>
    </div>
  </div>
</div>
`;
